internal/core/processor/universal/discoverer/discoverer_test.go

```diff
@@ -0,0 +1,305 @@
+			expectedError:     nil,
+		},
+		{
+			name: "error on storing universal contract",
+			event: scan.EventNewERC721Universal{
+				BaseURI:            "https://uloc.io/GlobalConsensus(3)/Parachain(9999)/AccountKey20(0x0000000000000000000000000000000000000000)/",
+				BlockNumber:        123,
+				NewContractAddress: common.HexToAddress("0xC3dd09D5387FA0Ab798e0ADC152d15b8d1a299DF"),
+			},
+			expectedContract: model.ERC721UniversalContract{
+				Address:           common.HexToAddress("0xc3dd09d5387fa0ab798e0adc152d15b8d1a299df"),
+				CollectionAddress: common.HexToAddress("0x0000000000000000000000000000000000000000"),
+				BlockNumber:       123,
+			},
+			errorOnStoringUniversalContract: fmt.Errorf("error on storing"),
+			expectedError:                   fmt.Errorf("error on storing"),
+		},
+		{
+			name: "error on get minted vents for universal contract",
+			event: scan.EventNewERC721Universal{
+				BaseURI:            "https://uloc.io/GlobalConsensus(3)/Parachain(9999)/AccountKey20(0x0000000000000000000000000000000000000000)/",
+				BlockNumber:        123,
+				NewContractAddress: common.HexToAddress("0xC3dd09D5387FA0Ab798e0ADC152d15b8d1a299DF"),
+			},
+			expectedContract: model.ERC721UniversalContract{
+				Address:           common.HexToAddress("0xc3dd09d5387fa0ab798e0adc152d15b8d1a299df"),
+				CollectionAddress: common.HexToAddress("0x0000000000000000000000000000000000000000"),
+				BlockNumber:       123,
+			},
+			errorOnGetMintedEvents: fmt.Errorf("error on getting minted events"),
+			expectedError:          fmt.Errorf("error on getting minted events"),
+		},
+
+		{
+			name: "successfully discover contract", // I am putting just success path since every other call just returns error
+			event: scan.EventNewERC721Universal{
+				BaseURI:            "https://uloc.io/GlobalConsensus(3)/Parachain(9999)/AccountKey20(0x0000000000000000000000000000000000000000)/",
+				BlockNumber:        123,
+				NewContractAddress: common.HexToAddress("0xC3dd09D5387FA0Ab798e0ADC152d15b8d1a299DF"),
+			},
+			expectedContract: model.ERC721UniversalContract{
+				Address:           common.HexToAddress("0xc3dd09d5387fa0ab798e0adc152d15b8d1a299df"),
+				CollectionAddress: common.HexToAddress("0x0000000000000000000000000000000000000000"),
+				BlockNumber:       123,
+			},
+			expectedError: nil,
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.TODO()
+			tx, client, scanner, validator := createMocks(t)
+
+			startingBlock := uint64(100)
+			lastBlock := uint64(200)
+
+			d := cDiscoverer.New(client, []string{}, scanner, validator)
+
+			// Mock the scanner's ScanNewUniversalEvents method
+			scanner.EXPECT().ScanNewUniversalEvents(ctx, big.NewInt(int64(startingBlock)), big.NewInt(int64(lastBlock))).
+				Return([]scan.EventNewERC721Universal{tt.event}, tt.errorOnScanning)
+
+			if tt.errorOnScanning == nil {
```

Comment:
This tests has too many ifs. We will struggle quite a lot to fix the tests if we need to refactor discoverer.
I suggest splitting the tests between error test cases and successful test cases. At least we will have less conditionals.

I'm ok even without table-tests and just `t.Run` subtests with repeated code

Reply:
ok